internal/pkg/agent/cmd/svc_update_linux.go

```diff
@@ -0,0 +1,61 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build linux
+
+package cmd
+
+import (
+	"fmt"
+	"os/exec"
+
+	"gopkg.in/ini.v1"
+
+	"github.com/kardianos/service"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+)
+
+func ensureServiceConfigUpToDate() error {
+	switch service.ChosenSystem().String() {
+	case "linux-systemd":
+		unitFilePath := "/etc/systemd/system/" + paths.ServiceName + ".service"
+		if err := ensureSystemdServiceConfigUpToDate(unitFilePath); err != nil {
+			return err
+		}
+
+		// Reload systemd unit configuration files
+		cmd := exec.Command("systemctl", "daemon-reload")
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("error reloading systemd unit configuration files: %w", err)
+		}
+
+	}
+
+	return nil
+}
+
+func ensureSystemdServiceConfigUpToDate(unitFilePath string) error {
+	// It is safe to use an INI file parser/writer for systemd unit files.
+	// See https://www.freedesktop.org/software/systemd/man/systemd.syntax.html
+	cfg, err := ini.Load(unitFilePath)
+	if err != nil {
+		return fmt.Errorf("error opening systemd unit file [%s]: %w", unitFilePath, err)
+	}
+
+	// Check if KillMode= is already present
+	if cfg.Section("Service").HasKey("KillMode") {
+		// Nothing more to do
+		return nil
+	}
+
+	// If KillMode= is not present, add it and set it to "process"
+	// See https://github.com/elastic/elastic-agent/pull/3220
+	cfg.Section("Service").Key("KillMode").SetValue("process")
+	if err := cfg.SaveTo(unitFilePath); err != nil {
+		return fmt.Errorf("error writing updated systemd unit file [%s]: %w", unitFilePath, err)
+	}
```

Comment:
I think we should NOTE that on a rollback this will not be changed back. The older Elastic Agent doesn't have a way to ensure that the service file is in its expected content.

I wonder how hard it would be to use the 3rd party `service` module that we use to render the file in memory and then just compare the contents of the file? That way we are not doing something specific here. It could just compare the 2 file contents and if different replace it with the new one and reload.

Is it not possible with the 3rd part `service` module to just have it overwrite the file instead of doing it here directly?

Reply:
> I wonder how hard it would be to use the 3rd party `service` module that we use to render the file in memory and then just compare the contents of the file? That way we are not doing something specific here. It could just compare the 2 file contents and if different replace it with the new one and reload.

I looked through https://pkg.go.dev/github.com/kardianos/service#section-documentation and I'm not seeing any APIs exposed by the `service` module that would help load the existing unit configuration file into memory (essentially a `service.Config` object) and compare it against our `service.Config` object.

> Is it not possible with the 3rd part `service` module to just have it overwrite the file instead of doing it here directly?

The reason I didn't go with the overwrite approach is because of a consideration that @cmacknz brought up in https://github.com/elastic/elastic-agent/pull/3220#issuecomment-1675311109: 

> The only other problem to consider is if users have manually edited parts of the configuration file, which we can likely handle by only changing the keys we care about and not replacing the entire file.

That's why I went with the approach of loading the current unit configuration file into memory, making a surgical edit if needed, and writing it back to disk.

Reply:
True, if they manually edited the file.